tests/testthat/test-qenv-within.R

```diff
@@ -0,0 +1,173 @@
+# styler: off
+# nolint start
+
+# evaluation ----
+## code acceptance ----
+testthat::test_that("simple expressions are evaluated", {
+  q <- new_qenv()
+  testthat::expect_no_error(within(q, 1 + 1))
+})
+
+testthat::test_that("compound expressions are evaluated", {
+  q <- new_qenv()
+  testthat::expect_no_error(
+    within(q, {
+      1 + 1
+    })
+  )
+  testthat::expect_no_error(
+    within(q, {
+      1 + 1
+      2 + 2
+    })
+  )
+  testthat::expect_no_error(
+    within(q, {
+      1 + 1; 2 + 2
+    })
+  )
+  testthat::expect_no_error(
+    within(q, {
+      1 +
+        1
+    })
+  )
+})
+
+testthat::test_that("characters in compound expressions are omitted", {
+  q <- new_qenv()
+  qq <- within(q, {
+    i <- iris
+    "1 + 1"
+    m <- mtcars
+  })
+  testthat::expect_identical(get_code(qq), c("i <- iris", "m <- mtcars"))
+})
+
+testthat::test_that("character-only compound expressions are ignored", {
+  q <- new_qenv()
+  qq <- within(q, {"1 + 1"})
+  testthat::expect_equal(q, qq)
+})
+
+testthat::test_that("symbol passed to `expr` raises warning", {
+  q <- new_qenv()
+  q <- within(q, i <- iris)
+  testthat::expect_warning(
+    qq <- within(q, i),
+    "Symbol passed to \"expr\""
+  )
+  testthat::expect_identical(get_code(qq), c("i <- iris", "i"))
+})
+
+
+## code identity ----
+testthat::test_that("differently formulated expressions yield the same code", {
+  q <- new_qenv()
+  q <- within(q, 1 + 1)
+  q <- within(q, {1 + 1})
+  q <- within(q, {
+    1 + 1
+  })
+  q <- within(q, {
+    1 +
+      1
+  })
+  all_code <- get_code(q)
+  testthat::expect_identical(
+    all_code,
+    rep("1 + 1", 4L)
+  )
+
+  q <- new_qenv()
+  q <- within(q, {1 + 1; 2 + 2})
+  q <- within(q, {
+    1 + 1; 2 + 2
+  })
+  q <- within(q, {
+    1 + 1
+    2 + 2
+  })
+  q <- within(q, {
+    1 + 1;
+    2 + 2
+  })
+  all_code <- get_code(q)
+  all_code_pairs <- lapply(seq_len(4L), function(x) all_code[((x - 1L) * 2L) + 1:2])
+  testthat::expect_identical(
+    all_code,
+    rep(c("1 + 1", "2 + 2"), 4L)
+  )
+})
+
+
+## injecting values ----
+testthat::test_that("external values can be injected into expressions through `...`", {
+  q <- new_qenv()
+
+  q <- within(q, {
+    i <- subset(iris, Species == "setosa")
+  })
+
+  q <- within(q, {
+    ii <- subset(iris, Species == species)
+  },
+  species = "virginica")
+
+  external_value <- "versicolor"
+  q <- within(q, {
+    iii <- subset(iris, Species == species)
+  },
+  species = external_value)
+
+  testthat::expect_identical(
+    get_code(q),
+    c(
+      "i <- subset(iris, Species == \"setosa\")",
+      "ii <- subset(iris, Species == \"virginica\")",
+      "iii <- subset(iris, Species == \"versicolor\")"
+    )
+  )
+
+  q <- new_qenv()
+  species <- "setosa"
+  qq <- within(q, {
+    i <- subset(iris, Species == species)
+  })
+  testthat::expect_s3_class(qq, "qenv.error")
```

Comment:
this should be a separate test - "failure in the expr evaluation causes within to return qenv.error"

Reply:
Sorry for above, I understand that meaning of this line is different. I propose separate testthat "within doesn't look for an object in the parent frame when not provided in `...`" - or similar